### PR TITLE
[MPS] Fix lintear for 5D tensors

### LINKED ISSUE #114942: [MPS] `nn.Linear` crashes if invoked for 5D tensors with assertion `New volume: xxx should match old volume: xxx'
### 🐛 Describe the bug

The complete code to reproduce the error:

```python
import numpy as np

import torch
import torchvision
import torch.nn as nn
import torch.optim as optim

class Model(nn.Module):
    
    def __init__(self):
        super().__init__()
        self.net = nn.Sequential(*[
            nn.Conv2d(64, 512, 1, stride=1, padding=0, groups=1),
        ])
        self.post = nn.Linear(8, 16)
        return
    
    def forward(self, xs):
        hs = self.net(xs)
        hs = hs.reshape([xs.shape[0], -1, 64, 64, 64]).transpose(1,4)
        hs = self.post(hs)
        return hs

device = "mps"

my_model = Model().to(device)
optimizer = optim.AdamW(my_model.parameters(), lr=1e-4)

batch_size = 2
target = torch.rand(batch_size, 64, 64, 64, 16).to(device)
xs = torch.rand(batch_size, 64, 64, 64).to(device)

pred = my_model(xs)
loss = torch.mean(target - pred)

optimizer.zero_grad()
loss.backward()
optimizer.step()

print(loss.item())
```

The output is as follows:
```
/AppleInternal/Library/BuildRoots/d9889869-120b-11ee-b796-7a03568b17ac/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSNDArray/Kernels/MPSNDArrayIdentity.mm:795: failed assertion `New volume: 4194304 should match old volume: 8388608 [reshapeWithCommandBuffer] MPSNDArrayIdentity.'
zsh: abort      python test.py
```

But if the 'device' is changed to 'cpu', the script can be run with no bug reported.

### Versions

PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.5.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.26.3
Libc version: N/A

Python version: 3.9.16 (main, Mar  8 2023, 04:29:44)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] onnxruntime==1.15.1
[pip3] open-clip-torch==2.18.0
[pip3] perlin-numpy==0.0.0
[pip3] pytorch-lightning==1.4.2
[pip3] pytorch-wavelets==1.3.0
[pip3] pytorch3d==0.7.4
[pip3] torch==2.1.0
[pip3] torchaudio==2.1.0
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.16.0
[pip3] torchviz==0.0.2
[conda] _pytorch_select           0.1                       cpu_0
[conda] blas                      1.0                         mkl
[conda] ffmpeg                    4.3                  h0a44026_0    pytorch
[conda] libmklml                  2019.0.5                      0
[conda] mkl                       2019.4                      233
[conda] mkl-service               2.3.0            py39h9ed2024_0
[conda] mkl_fft                   1.3.0            py39ha059aab_0
[conda] mkl_random                1.0.2            py39h16bde0e_0
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] open-clip-torch           2.18.0                   pypi_0    pypi
[conda] perlin-numpy              0.0.0                    pypi_0    pypi
[conda] pytorch-lightning         1.4.2                    pypi_0    pypi
[conda] pytorch-wavelets          1.3.0                    pypi_0    pypi
[conda] pytorch3d                 0.7.4                    pypi_0    pypi
[conda] torch                     2.1.0                    pypi_0    pypi
[conda] torchaudio                2.1.0                    pypi_0    pypi
[conda] torchmetrics              0.6.0                    pypi_0    pypi
[conda] torchvision               0.16.0                   pypi_0    pypi
[conda] torchviz                  0.0.2                    pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr

### PR DESCRIPTION
**Pull request by malfet**
torch.nn.Linear crashes with internal assert if invoked with 5D tensors,
due to the bug in MPS framework, i.e. invoking
```swift
import MetalPerformanceShadersGraph

let graph = MPSGraph()
let x = graph.constant(1, shape: [2, 1, 2, 1, 2], dataType: .float32)
let y = graph.constant(1, shape: [2, 3], dataType: .float32)
let z = graph.matrixMultiplication(primary: x, secondary: y, name: nil)
let device = MTLCreateSystemDefaultDevice()!
let buf = device.makeBuffer(length: 48)!
let td = MPSGraphTensorData(buf, shape: [2, 1, 2, 1, 3], dataType: .int32)
let cmdBuf = MPSCommandBuffer(from: device.makeCommandQueue()!)
graph.encode(to: cmdBuf, feeds: [:], targetOperations: nil, resultsDictionary: [z:td], executionDescriptor: nil)
cmdBuf.commit()
```
crashes with
```
AppleInternal/Library/BuildRoots/0032d1ee-80fd-11ee-8227-6aecfccc70fe/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSNDArray/Kernels/MPSNDArrayIdentity.mm:813: failed assertion `New volume: 4 should match old volume: 8 [reshapeWithCommandBuffer] MPSNDArrayIdentity.'
zsh: abort      ./build/matmul
```

Workaround the issue by flattening the forward and backward tensors if number of dimentions is greater than 4

Add regression tests to Linear opinfo samples

Fixes https://github.com/pytorch/pytorch/issues/114942

### CODE REVIEW

**Comment by malfet** [comment]
Android failures are unrelated, just don't want to rebase...

**Review by janeyx99 — APPROVED** [review]
This is a better state than before, but I have questions about the longterm:
- Is the long term to fix it from the MPS side and remove this reshape?
- Is the test_neg_view/test_conj_view failure expected?

**Comment by malfet** [comment]
> * Is the long term to fix it from the MPS side and remove this reshape?

Yes, though perhaps reshape can stay here if it does not affect the perf (as far as I understand from the comment to https://github.com/pytorch/pytorch/pull/91114 reshape actually makes it faster for 3+ dim tensors)

> * Is the test_neg_view/test_conj_view failure expected?

I didn't expect it, so I filed https://github.com/pytorch/pytorch/issues/117854 to investigate what is going on (looks like a bug in CPU implementation of the backward operation, likely deep in mkldnn)
